<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Introduction · ProRF.jl</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href><img src="assets/logo.svg" alt="ProRF.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href>ProRF.jl</a></span></div><form class="docs-search" action="search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li class="is-active"><a class="tocitem" href>Introduction</a><ul class="internal"><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li><a class="tocitem" href="#Install"><span>Install</span></a></li><li class="toplevel"><a class="tocitem" href="#Molecular-mass-of-amino-acid"><span>Molecular mass of amino acid</span></a></li></ul></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="library/datapro/">Data preprocess</a></li><li><a class="tocitem" href="library/ranfor/">Random Forest</a></li><li><a class="tocitem" href="library/other/">Toolbox</a></li><li><a class="tocitem" href="library/findex/">Index</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>Introduction</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Introduction</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Chemical118/ProRF.jl/blob/master/docs/src/index.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="ProRF.jl-Documentation"><a class="docs-heading-anchor" href="#ProRF.jl-Documentation">ProRF.jl Documentation</a><a id="ProRF.jl-Documentation-1"></a><a class="docs-heading-anchor-permalink" href="#ProRF.jl-Documentation" title="Permalink"></a></h1><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><p><code>ProRF</code> provides a full process for applying the random forest model of protein sequences using <code>DecisionTree</code>.</p><h2 id="Install"><a class="docs-heading-anchor" href="#Install">Install</a><a id="Install-1"></a><a class="docs-heading-anchor-permalink" href="#Install" title="Permalink"></a></h2><p>&lt;!– !!! warning     <code>ProRF</code> uses Python module <code>Bokeh</code>, <code>Matplotlib</code> to provide UI. Please install these module or execute below code before add <code>ProRF</code>.</p><pre><code class="language-bash hljs">    $ pip install matplotlib
    $ pip install bokeh
    ```
    For more information, see `PyCall` [main page](https://github.com/JuliaPy/PyCall.jl). --&gt;
</code></pre><p>julia using Pkg Pkg.add(url=&quot;https://github.com/Chemical118/ProRF.jl&quot;)</p><pre><code class="nohighlight hljs">
## Examples
!!! tip &quot;Performance Tip&quot;
    `ProRF` support parallel computing, please turn on the julia with multiple threads. This can speed up execution time fairly.

    ```bash
    $ julia --threads auto
    ```
    For more information, read Multi-Threading [documentaion](https://docs.julialang.org/en/v1/manual/multi-threading/).

!!! note
    `ProRF` recommends interactive mode like `IJulia`. If you want to run in non-interactive mode, execute below code to see graphs. However, `ProRF` doesn&#39;t guarantee that you can see graphs.
    ```julia
    using ProRF
    julia_isinteractive(false)
    ```

### Data preprocessing
`ProRF` has a useful function for preprocessing data.</code></pre><p>julia using ProRF, Printf</p><p>Find, Lind = data<em>preprocess</em>index(&quot;Data/algpdata.fasta&quot;, val_mode=true) @printf &quot;%d %d\n&quot; Find Lind</p><p>data<em>preprocess</em>fill(Find, Lind,                      &quot;Data/algpdata.fasta&quot;,                      &quot;Data/Mega/ealtreedata.nwk&quot;,                      &quot;Data/jealgpdata.fasta&quot;,                      val_mode=true);</p><p>view_sequence(&quot;Data/jealgpdata.fasta&quot;, save=true)</p><pre><code class="nohighlight hljs">
### Find best random forest arguments
`ProRF` helps you find arguments for the random forest.</code></pre><p>julia using ProRF, Printf</p><p>RI = RFI(&quot;Data/jealgpdata.fasta&quot;, &quot;Data/data.xlsx&quot;, 2:1:10, 100:10:500) X, Y, L = get_data(RI, 2, &#39;D&#39;)</p><p>MeZ, SdZ = iter<em>get</em>reg<em>value(RI, X, Y, 10, val</em>mode=true)</p><p>view<em>reg3d(RI, MeZ, title=&quot;NRMSE value&quot;, azim=90, scale=3) view</em>reg3d(RI, SdZ, title=&quot;NRMSE SD value&quot;, elev=120, scale=3)</p><p>N<em>Feature, N</em>Tree = get<em>reg</em>value<em>loc(RI, MZ) @printf &quot;Best Arguments : %d %d\n&quot; N</em>Feature N_Tree</p><pre><code class="nohighlight hljs">
### Execute random forest
`ProRF` executes random forest flexibly and easily.</code></pre><p>julia using ProRF, Printf</p><h1 id="Molecular-mass-of-amino-acid"><a class="docs-heading-anchor" href="#Molecular-mass-of-amino-acid">Molecular mass of amino acid</a><a id="Molecular-mass-of-amino-acid-1"></a><a class="docs-heading-anchor-permalink" href="#Molecular-mass-of-amino-acid" title="Permalink"></a></h1><p>myDict = Dict(&#39;A&#39; =&gt; 89, &#39;R&#39; =&gt; 174, &#39;N&#39; =&gt; 132, &#39;D&#39; =&gt; 133, &#39;C&#39; =&gt; 121, &#39;Q&#39; =&gt; 146, &#39;E&#39; =&gt; 147, &#39;G&#39; =&gt; 75, &#39;H&#39; =&gt; 155, &#39;I&#39; =&gt; 131, &#39;L&#39; =&gt; 131, &#39;K&#39; =&gt; 146, &#39;M&#39; =&gt; 149,  &#39;F&#39; =&gt; 165, &#39;P&#39; =&gt; 115, &#39;S&#39; =&gt; 105, &#39;T&#39; =&gt; 119, &#39;W&#39; =&gt; 204, &#39;Y&#39; =&gt; 181, &#39;V&#39; =&gt; 117)</p><p>R = RF(&quot;Data/jealgpdata.fasta&quot;, &quot;Data/data.xlsx&quot;) X, Y, L = get_data(R, 2, &#39;D&#39;, convert=myDict)</p><p>M = rf<em>model(X, Y, N</em>Feature, N_Tree) @printf &quot;Total NRMSE : %.6f\n&quot; nrmse(M, X, Y)</p><p>MeF, SdF = iter<em>get</em>reg<em>importance(R, X, Y, L, N</em>Feature, N<em>Tree, 100, val</em>mode=true) view_importance(R, L, MeF, SdF)</p><p>for (fe, loc) in sort(collect(zip(MeF, get<em>amino</em>loc(R, L))), by = x -&gt; x[1])[1:10]     @printf &quot;Location %s : %.4f\n&quot; loc fe end ```</p></article><nav class="docs-footer"><a class="docs-footer-nextpage" href="library/datapro/">Data preprocess »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.22 on <span class="colophon-date" title="Tuesday 26 July 2022 07:44">Tuesday 26 July 2022</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
