<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Toolbox · ProRF.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.svg" alt="ProRF.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">ProRF.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Introduction</a></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../datapro/">Data preprocess</a></li><li><a class="tocitem" href="../ranfor/">Random Forest</a></li><li class="is-active"><a class="tocitem" href>Toolbox</a><ul class="internal"><li><a class="tocitem" href="#Random-Forest"><span>Random Forest</span></a></li><li><a class="tocitem" href="#Data-preprocessing"><span>Data preprocessing</span></a></li><li><a class="tocitem" href="#Convert-dictionary"><span>Convert dictionary</span></a></li><li><a class="tocitem" href="#Others"><span>Others</span></a></li><li><a class="tocitem" href="#Reference"><span>Reference</span></a></li></ul></li><li><a class="tocitem" href="../findex/">Index</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Toolbox</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Toolbox</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Chemical118/ProRF.jl/blob/master/docs/src/library/other.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Toolbox"><a class="docs-heading-anchor" href="#Toolbox">Toolbox</a><a id="Toolbox-1"></a><a class="docs-heading-anchor-permalink" href="#Toolbox" title="Permalink"></a></h1><p>Documentaion for <code>ProRF</code>&#39;s useful things to perform data preprocessing or random forest.</p><h2 id="Random-Forest"><a class="docs-heading-anchor" href="#Random-Forest">Random Forest</a><a id="Random-Forest-1"></a><a class="docs-heading-anchor-permalink" href="#Random-Forest" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ProRF.get_rf_value" href="#ProRF.get_rf_value"><code>ProRF.get_rf_value</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">get_rf_value(X::Matrix{Float64}, Y::Vector{Float64};
             val_mode::Bool=false,
             iter::Int=10, test_size::Float64=0.3,
             feat_range::Int=4, base_tree::Int=50,
             memory_usage::Real=4.0,
             max_tree::Int=1000)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; NFeat, NTree, NDepth = get_rf_value(X, Y, iter=5, memory_usage=10, val_mode=true);</code></pre><p>Find best three arguments for random forest.</p><p><strong>Arguments</strong></p><ul><li><code>X::Matrix{Float64}</code> : <code>X</code> data.</li><li><code>Y::Vector{Float64}</code> : <code>Y</code> data.</li><li><code>val_mode::Bool</code> : when <code>val_mode</code> is true, function don&#39;t display anything.</li><li><code>iter::Int</code> : number of operations iterations.</li><li><code>test_size::Float64</code> : size of test set.</li><li><code>feat_range::Int</code> : scope of search for number of selected features.</li><li><code>base_tree::Int</code> : number of trees used when navigating.</li><li><code>memory_usage::Real</code> : available memory capacity (GB)</li><li><code>max_tree::Int</code> : thresholds fornumber of trees for performance.</li></ul><p><strong>Return</strong></p><ul><li><code>opt_feat::Int</code> : optimized number of selected features.</li><li><code>opt_tree::Int</code> : optimized number of trees.</li><li><code>opt_depth::Int</code> : optimized maximum depth of the tree.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L1480-L1509">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.train_test_split" href="#ProRF.train_test_split"><code>ProRF.train_test_split</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">train_test_split(X::Matrix{Float64}, Y::Vector{Float64};
                 test_size::Float64=0.3, 
                 data_state::UInt64=@seed)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2);</code></pre><p>Split <code>X</code>, <code>Y</code> data to train, test set.</p><p><strong>Arguments</strong></p><ul><li><code>X::Matrix{Float64}</code> : <code>X</code> data.</li><li><code>Y::Vector{Float64}</code> : <code>Y</code> data.</li><li><code>test_size::Float64</code> : size of test set.</li><li><code>data_state::UInt64</code> : seed used to split data.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L331-L348">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.nrmse" href="#ProRF.nrmse"><code>ProRF.nrmse</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nrmse(pre::Vector{Float64}, tru::Vector{Float64})</code></pre><p>Compute normalized root mean square error with predict value and true value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L216-L220">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.test_nrmse" href="#ProRF.test_nrmse"><code>ProRF.test_nrmse</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">test_nrmse(regr::RandomForestRegressor, X::Matrix{Float64}, Y::Vector{Float64})</code></pre><p>Compute normalized root mean square error with regression model, <code>X</code> and <code>Y</code> data.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L225-L229">source</a></section><section><div><pre><code class="nohighlight hljs">test_nrmse(regr::RandomForestRegressor, X::Matrix{Float64}, Y::Vector{Float64},
           data_state::UInt64;
           test_size::Float64=0.3, test_mode::Bool=true)</code></pre><p>Compute test or train set normalized root mean square error with regression model, <code>X</code>, <code>Y</code> data and seed.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L234-L240">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.min_max_norm" href="#ProRF.min_max_norm"><code>ProRF.min_max_norm</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">min_max_norm(data::Vector{Float64})</code></pre><p>Min-max normalization function.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L786-L790">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.save_model" href="#ProRF.save_model"><code>ProRF.save_model</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">save_model(model_loc::String, regr::RandomForestRegressor)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; M = rf_model(X, Y, 5, 300, val_mode=true);

julia&gt; save_model(&quot;model.jld&quot;, M);
ERROR: model.jld is not .jld2 file

julia&gt; save_model(&quot;model.jld2&quot;, M);</code></pre><p>Save <code>RandomForestRegressor</code> model using <code>JLD2</code>, make sure filename extension set to <code>.jld2</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L358-L371">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.load_model" href="#ProRF.load_model"><code>ProRF.load_model</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">load_model(model_loc::String)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; M = load_model(&quot;model.jld2&quot;);

julia&gt; X, Y, L = get_data(R, 9, &#39;E&#39;);

julia&gt; @printf &quot;Total NRMSE : %.6f\n&quot; test_nrmse(M, X, Y)
Total NRMSE : 0.136494</code></pre><p>Load <code>RandomForestRegressor</code> model using <code>JLD2</code>, make sure filename extension set to <code>.jld2</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L379-L392">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.parallel_predict" href="#ProRF.parallel_predict"><code>ProRF.parallel_predict</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">parallel_predict(regr::RandomForestRegressor, X::Matrix{Float64})</code></pre><p>Execute <code>DecisionTree.predict(regr, X)</code> in parallel.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L250-L254">source</a></section><section><div><pre><code class="nohighlight hljs">parallel_predict(regr::RandomForestRegressor, L::Vector{Int},
                 seq_vector::Vector{String};
                 convert::Dict{Char, Float64}=ProRF.volume)</code></pre><p>Get raw sequence vector and <code>L</code> data to make <code>X</code> data and execute <code>DecisionTree.predict(regr, X)</code> in parallel.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L259-L265">source</a></section></article><h2 id="Data-preprocessing"><a class="docs-heading-anchor" href="#Data-preprocessing">Data preprocessing</a><a id="Data-preprocessing-1"></a><a class="docs-heading-anchor-permalink" href="#Data-preprocessing" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ProRF.view_sequence" href="#ProRF.view_sequence"><code>ProRF.view_sequence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">view_sequence(fasta_loc::String, amino_loc::Int=1;
              fontsize::Int=9,
              seq_width::Int=800,
              save::Bool=false)</code></pre><p>Display <code>.fasta</code> file sequence using <code>PyCall</code>.</p><p>Made it by referring to <a href="https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner">bokeh sequence aligner visualization program</a>.</p><p><strong>Arguments</strong></p><ul><li><code>fasta_loc::String</code> : location of <code>.fasta</code> file.</li><li><code>amino_loc::Int</code> : start index for <code>.fasta</code> file (when value is not determined, set to 1).</li><li><code>fontsize::Int</code> : font size of sequence.</li><li><code>seq_width::Int</code> : gap width between sequences.</li><li><code>save::Bool</code> : save <code>.html</code> viewer.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L272-L288">source</a></section><section><div><pre><code class="nohighlight hljs">view_sequence(R::AbstractRF;
              fontsize::Int=9,
              seq_width::Int=800,
              save::Bool=false)</code></pre><p>Display <code>.fasta</code> file sequence exists at a location in the <a href="../ranfor/#ProRF.AbstractRF"><code>AbstractRF</code></a> object using <code>PyCall</code>.</p><p>Made it by referring to <a href="https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner">bokeh sequence aligner visualization program</a>.</p><p><strong>Arguments</strong></p><ul><li><code>R::AbstractRF</code> : for both <a href="../ranfor/#ProRF.RF"><code>RF</code></a> and <a href="../ranfor/#ProRF.RFI"><code>RFI</code></a>.</li><li><code>fontsize::Int</code> : font size of sequence.</li><li><code>seq_width::Int</code> : gap width between sequences.</li><li><code>save::Bool</code> : save <code>.html</code> viewer.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L300-L315">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.view_mutation" href="#ProRF.view_mutation"><code>ProRF.view_mutation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">view_mutation(fasta_loc::String)</code></pre><p>Analyze data from <code>.fasta</code> file location of AbstractRF, then draw histogram, line graph about the mutation distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L168-L172">source</a></section><section><div><pre><code class="nohighlight hljs">view_mutation(R::AbstractRF)</code></pre><p>Analyze data from <code>.fasta</code> file, then draw histogram, line graph about the mutation distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L177-L181">source</a></section></article><h2 id="Convert-dictionary"><a class="docs-heading-anchor" href="#Convert-dictionary">Convert dictionary</a><a id="Convert-dictionary-1"></a><a class="docs-heading-anchor-permalink" href="#Convert-dictionary" title="Permalink"></a></h2><p>Convert dictionary for <a href="../ranfor/#ProRF.get_data"><code>get_data</code></a></p><table><tr><th style="text-align: left">Amino acid</th><th style="text-align: left">Molar volume [<a href="#Reference">4</a>]</th><th style="text-align: left">pI [<a href="#Reference">2</a>]</th><th style="text-align: left">Hydrophobicity [<a href="#Reference">3</a>]</th><th style="text-align: left">Frequency [<a href="#Reference">1</a>]</th></tr><tr><td style="text-align: left">Alanine</td><td style="text-align: left">88.6</td><td style="text-align: left">6.00</td><td style="text-align: left">51</td><td style="text-align: left">0.0777</td></tr><tr><td style="text-align: left">Arginine</td><td style="text-align: left">173.4</td><td style="text-align: left">10.76</td><td style="text-align: left">-144</td><td style="text-align: left">0.0627</td></tr><tr><td style="text-align: left">Asparagine</td><td style="text-align: left">114.1</td><td style="text-align: left">5.41</td><td style="text-align: left">-84</td><td style="text-align: left">0.0336</td></tr><tr><td style="text-align: left">Aspartic acid</td><td style="text-align: left">111.1</td><td style="text-align: left">2.77</td><td style="text-align: left">-78</td><td style="text-align: left">0.0542</td></tr><tr><td style="text-align: left">Cysteine</td><td style="text-align: left">108.5</td><td style="text-align: left">5.07</td><td style="text-align: left">137</td><td style="text-align: left">0.0078</td></tr><tr><td style="text-align: left">Glutamic acid</td><td style="text-align: left">138.4</td><td style="text-align: left">3.22</td><td style="text-align: left">-115</td><td style="text-align: left">0.0859</td></tr><tr><td style="text-align: left">Glutamine</td><td style="text-align: left">143.8</td><td style="text-align: left">5.65</td><td style="text-align: left">-128</td><td style="text-align: left">0.0315</td></tr><tr><td style="text-align: left">Glycine</td><td style="text-align: left">60.1</td><td style="text-align: left">5.97</td><td style="text-align: left">-13</td><td style="text-align: left">0.0730</td></tr><tr><td style="text-align: left">Histidine</td><td style="text-align: left">153.2</td><td style="text-align: left">7.59</td><td style="text-align: left">-55</td><td style="text-align: left">0.0192</td></tr><tr><td style="text-align: left">Isoleucine</td><td style="text-align: left">166.7</td><td style="text-align: left">6.02</td><td style="text-align: left">106</td><td style="text-align: left">0.0666</td></tr><tr><td style="text-align: left">Leucine</td><td style="text-align: left">166.7</td><td style="text-align: left">5.98</td><td style="text-align: left">103</td><td style="text-align: left">0.0891</td></tr><tr><td style="text-align: left">Lysine</td><td style="text-align: left">168.6</td><td style="text-align: left">9.74</td><td style="text-align: left">-205</td><td style="text-align: left">0.0776</td></tr><tr><td style="text-align: left">Methionine</td><td style="text-align: left">162.9</td><td style="text-align: left">5.74</td><td style="text-align: left">73</td><td style="text-align: left">0.0241</td></tr><tr><td style="text-align: left">Phenylalanine</td><td style="text-align: left">189.9</td><td style="text-align: left">5.48</td><td style="text-align: left">108</td><td style="text-align: left">0.0361</td></tr><tr><td style="text-align: left">Proline</td><td style="text-align: left">112.7</td><td style="text-align: left">6.30</td><td style="text-align: left">-79</td><td style="text-align: left">0.0435</td></tr><tr><td style="text-align: left">Serine</td><td style="text-align: left">89.0</td><td style="text-align: left">5.68</td><td style="text-align: left">-26</td><td style="text-align: left">0.0466</td></tr><tr><td style="text-align: left">Threonine</td><td style="text-align: left">116.1</td><td style="text-align: left">5.60</td><td style="text-align: left">-3</td><td style="text-align: left">0.0487</td></tr><tr><td style="text-align: left">Tryptophan</td><td style="text-align: left">227.8</td><td style="text-align: left">5.89</td><td style="text-align: left">69</td><td style="text-align: left">0.0102</td></tr><tr><td style="text-align: left">Tyrosine</td><td style="text-align: left">193.6</td><td style="text-align: left">5.66</td><td style="text-align: left">11</td><td style="text-align: left">0.0300</td></tr><tr><td style="text-align: left">Valine</td><td style="text-align: left">140.0</td><td style="text-align: left">5.96</td><td style="text-align: left">108</td><td style="text-align: left">0.0817</td></tr></table><p>Frequency of amino acid is used to estimate the value of unknown amino acid.</p><article class="docstring"><header><a class="docstring-binding" id="ProRF.volume" href="#ProRF.volume"><code>ProRF.volume</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>Convert dictionary about molar volume of amino acid.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L1576-L1578">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.pI" href="#ProRF.pI"><code>ProRF.pI</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>Convert dictionary about pI of amino acid.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L1582-L1584">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.hydrophobicity" href="#ProRF.hydrophobicity"><code>ProRF.hydrophobicity</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>Convert dictionary about hydrophobicity of amino acid.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L1588-L1590">source</a></section></article><h2 id="Others"><a class="docs-heading-anchor" href="#Others">Others</a><a id="Others-1"></a><a class="docs-heading-anchor-permalink" href="#Others" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ProRF._julia_interactive" href="#ProRF._julia_interactive"><code>ProRF._julia_interactive</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>Check julia currently running is interactive or not.</p><p>You can change the value in <a href="#ProRF.julia_isinteractive"><code>julia_isinteractive</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L15-L19">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.julia_isinteractive" href="#ProRF.julia_isinteractive"><code>ProRF.julia_isinteractive</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">julia_isinteractive()</code></pre><p>Check <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L401-L405">source</a></section><section><div><pre><code class="nohighlight hljs">julia_isinteractive(x::Bool)</code></pre><p>Set <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L410-L414">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.@seed" href="#ProRF.@seed"><code>ProRF.@seed</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>Return <code>UInt64</code> range integer <code>MersenneTwister</code> RNG object seed. keep in mind that when macro executed, the seed is initialized.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L140-L142">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.@show_pyplot" href="#ProRF.@show_pyplot"><code>ProRF.@show_pyplot</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>When <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> is on, execute <code>display(gcf())</code> or <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> is off, execute <code>show()</code> and wait until the user inputs enter.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/0c89abf1e2339bf1e9e9d8e02f3dc4ec13d108e9/src/ProRF.jl#L147-L149">source</a></section></article><h2 id="Reference"><a class="docs-heading-anchor" href="#Reference">Reference</a><a id="Reference-1"></a><a class="docs-heading-anchor-permalink" href="#Reference" title="Permalink"></a></h2><p>[1] : Brooks, D. J., Fresco, J. R., Lesk, A. M., &amp; Singh, M. (2002). Evolution of amino acid frequencies in proteins over deep time: inferred order of introduction of amino acids into the genetic code. <em>Molecular Biology and Evolution, 19</em>(10), 1645–1655. <a href="https://doi.org/10.1093/oxfordjournals.molbev.a003988">https://doi.org/10.1093/oxfordjournals.molbev.a003988</a></p><p>[2] :  Lide, D. R. (Ed.). (1991). <em>Hdbk of chemistry &amp; physics 72nd edition</em> (72nd ed.). CRC Press.</p><p>[3] :  Naderi-Manesh, H., Sadeghi, M., Arab, S., &amp; Moosavi Movahedi, A. A. (2001). Prediction of protein surface accessibility with information theory. <em>Proteins, 42</em>(4), 452–459. <a href="https://doi.org/10.1002/1097-0134(20010301)42:4&lt;452::aid-prot40&gt;3.0.co;2-q">https://doi.org/10.1002/1097-0134(20010301)42:4&lt;452::aid-prot40&gt;3.0.co;2-q</a></p><p>[4] :  Zamyatnin, A. (1984). Amino Acid, Peptide, and Protein Volume in Solution. <em>Annual Review of Biophysics and Biomolecular Structure, 13</em>(1), 145–165. <a href="https://doi.org/10.1146/annurev.biophys.13.1.145">https://doi.org/10.1146/annurev.biophys.13.1.145</a></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../ranfor/">« Random Forest</a><a class="docs-footer-nextpage" href="../findex/">Index »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.22 on <span class="colophon-date" title="Friday 29 July 2022 09:28">Friday 29 July 2022</span>. Using Julia version 1.6.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
