<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Toolbox · ProRF.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.044/juliamono.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">ProRF.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Introduction</a></li><li><span class="tocitem">Library</span><ul><li><a class="tocitem" href="../datapro/">Data preprocess</a></li><li><a class="tocitem" href="../ranfor/">Random Forest</a></li><li class="is-active"><a class="tocitem" href>Toolbox</a><ul class="internal"><li><a class="tocitem" href="#Random-Forest"><span>Random Forest</span></a></li><li><a class="tocitem" href="#Data-preprocessing"><span>Data preprocessing</span></a></li><li><a class="tocitem" href="#Others"><span>Others</span></a></li></ul></li><li><a class="tocitem" href="../findex/">Index</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Library</a></li><li class="is-active"><a href>Toolbox</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Toolbox</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Chemical118/ProRF.jl/blob/master/docs/src/library/other.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Toolbox-functions"><a class="docs-heading-anchor" href="#Toolbox-functions">Toolbox functions</a><a id="Toolbox-functions-1"></a><a class="docs-heading-anchor-permalink" href="#Toolbox-functions" title="Permalink"></a></h1><p>Documentaion for <code>ProRF</code>&#39;s useful function to perform data preprocessing or random forest.</p><h2 id="Random-Forest"><a class="docs-heading-anchor" href="#Random-Forest">Random Forest</a><a id="Random-Forest-1"></a><a class="docs-heading-anchor-permalink" href="#Random-Forest" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ProRF.blosum_matrix" href="#ProRF.blosum_matrix"><code>ProRF.blosum_matrix</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">blosum_matrix(num::Int)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; blosum_matrix(90)[&#39;A&#39;, &#39;A&#39;]
5

julia&gt; blosum_matrix(82)
ERROR: There are no Matrix such as BLOSUM82</code></pre><p>Return correct <code>BioAlignments</code> BLOSUM matrix.</p><p>Possible matrices are BLOSUM45, BLOSUM50, BLOSUM62, BLOSUM80, BLOSUM90.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L122-L136">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.train_test_split" href="#ProRF.train_test_split"><code>ProRF.train_test_split</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">train_test_split(X::Matrix{Float64}, Y::Vector{Float64};
                 test_size::Float64=0.3, 
                 data_state::UInt64=@seed_u64)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; x_train, x_test, y_train, y_test = train_test_split(X, Y, test_size=0.2);</code></pre><p>Split <code>X</code>, <code>Y</code> data to train, test set.</p><p><strong>Arguments</strong></p><ul><li><code>X::Matrix{Float64}</code> : <code>X</code> data.</li><li><code>Y::Vector{Float64}</code> : <code>Y</code> data.</li><li><code>test_size::Float64</code> : size of test set.</li><li><code>data_state::UInt64</code> : seed used to split data.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L311-L328">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.nrmse" href="#ProRF.nrmse"><code>ProRF.nrmse</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">nrmse(pre::Vector{Float64}, tru::Vector{Float64})</code></pre><p>Compute normalized root mean square error with predict value and true value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L201-L205">source</a></section><section><div><pre><code class="nohighlight hljs">nrmse(regr::RandomForestRegressor, X::Matrix{Float64}, tru::Vector{Float64})</code></pre><p>Compute normalized root mean square error with regression model, <code>X</code> data and true value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L210-L214">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.save_model" href="#ProRF.save_model"><code>ProRF.save_model</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">save_model(model_loc::String, regr::RandomForestRegressor)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; M = rf_model(X, Y, 5, 300, val_mode=true);

julia&gt; save_model(&quot;model.jld&quot;, M);
ERROR: model.jld is not .jld2 file

julia&gt; save_model(&quot;model.jld2&quot;, M);</code></pre><p>Save <code>RandomForestRegressor</code> model using <code>JLD2</code>, make sure filename extension set to <code>.jld2</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L338-L351">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.load_model" href="#ProRF.load_model"><code>ProRF.load_model</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">load_model(model_loc::String)</code></pre><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; M = load_model(&quot;model.jld2&quot;);

julia&gt; X, Y, L = get_data(R, 9, &#39;E&#39;);

julia&gt; @printf &quot;Total NRMSE : %.6f\n&quot; nrmse(M, X, Y)
Total NRMSE : 0.136494</code></pre><p>Load <code>RandomForestRegressor</code> model using <code>JLD2</code>, make sure filename extension set to <code>.jld2</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L359-L372">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.parallel_predict" href="#ProRF.parallel_predict"><code>ProRF.parallel_predict</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">parallel_predict(regr::RandomForestRegressor, X::Matrix{Float64})</code></pre><p>Execute <code>DecisionTree.predict(regr, X)</code> in parallel.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L219-L223">source</a></section><section><div><pre><code class="nohighlight hljs">parallel_predict(regr::RandomForestRegressor, X::Matrix{Float64})</code></pre><p>Get raw sequence vector and <code>L</code> data to make <code>X</code> data and execute <code>DecisionTree.predict(regr, X)</code> in parallel.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L234-L238">source</a></section></article><h2 id="Data-preprocessing"><a class="docs-heading-anchor" href="#Data-preprocessing">Data preprocessing</a><a id="Data-preprocessing-1"></a><a class="docs-heading-anchor-permalink" href="#Data-preprocessing" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ProRF.view_sequence" href="#ProRF.view_sequence"><code>ProRF.view_sequence</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">view_sequence(fasta_loc::String, amino_loc::Int=1;
              fontsize::Int=9,
              seq_width::Int=800,
              save::Bool=false)</code></pre><p>Display <code>.fasta</code> file sequence using <code>PyCall</code>.</p><p>Made it by referring to <a href="https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner">bokeh sequence aligner visualization program</a>.</p><p><strong>Arguments</strong></p><ul><li><code>fasta_loc::String</code> : location of <code>.fasta</code> file.</li><li><code>amino_loc::Int</code> : start index for <code>.fasta</code> file (when value is not determined, set to 1).</li><li><code>fontsize::Int</code> : font size of sequence.</li><li><code>seq_width::Int</code> : gap width between sequences.</li><li><code>save::Bool</code> : save <code>.html</code> viewer.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L252-L268">source</a></section><section><div><pre><code class="nohighlight hljs">view_sequence(R::AbstractRF;
              fontsize::Int=9,
              seq_width::Int=800,
              save::Bool=false)</code></pre><p>Display <code>.fasta</code> file sequence exists at a location in the <a href="../ranfor/#ProRF.AbstractRF"><code>AbstractRF</code></a> object using <code>PyCall</code>.</p><p>Made it by referring to <a href="https://dmnfarrell.github.io/bioinformatics/bokeh-sequence-aligner">bokeh sequence aligner visualization program</a>.</p><p><strong>Arguments</strong></p><ul><li><code>R::AbstractRF</code> : for both <a href="../ranfor/#ProRF.RF"><code>RF</code></a> and <a href="../ranfor/#ProRF.RFI"><code>RFI</code></a>.</li><li><code>fontsize::Int</code> : font size of sequence.</li><li><code>seq_width::Int</code> : gap width between sequences.</li><li><code>save::Bool</code> : save <code>.html</code> viewer.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L280-L295">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.view_mutation" href="#ProRF.view_mutation"><code>ProRF.view_mutation</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">view_mutation(fasta_loc::String)</code></pre><p>Analyze data from <code>.fasta</code> file location of AbstractRF, then draw histogram, line graph about the mutation distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L153-L157">source</a></section><section><div><pre><code class="nohighlight hljs">view_mutation(R::AbstractRF)</code></pre><p>Analyze data from <code>.fasta</code> file, then draw histogram, line graph about the mutation distribution.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L162-L166">source</a></section></article><h2 id="Others"><a class="docs-heading-anchor" href="#Others">Others</a><a id="Others-1"></a><a class="docs-heading-anchor-permalink" href="#Others" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-binding" id="ProRF._julia_interactive" href="#ProRF._julia_interactive"><code>ProRF._julia_interactive</code></a> — <span class="docstring-category">Constant</span></header><section><div><p>Check julia currently running is interactive or not.</p><p>You can change the value in <a href="#ProRF.julia_isinteractive"><code>julia_isinteractive</code></a>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L14-L18">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.julia_isinteractive" href="#ProRF.julia_isinteractive"><code>ProRF.julia_isinteractive</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">julia_isinteractive()</code></pre><p>Check <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L381-L385">source</a></section><section><div><pre><code class="nohighlight hljs">julia_isinteractive(x::Bool)</code></pre><p>Set <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> value.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L390-L394">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.@seed_i" href="#ProRF.@seed_i"><code>ProRF.@seed_i</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>Return non-negative <code>Int</code> range integer <code>MersenneTwister</code> RNG object seed. keep in mind that when macro executed, the seed is initialized.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L92-L94">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.@seed_u64" href="#ProRF.@seed_u64"><code>ProRF.@seed_u64</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>Return <code>UInt64</code> range integer <code>MersenneTwister</code> RNG object seed. keep in mind that when macro executed, the seed is initialized.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L99-L101">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="ProRF.@show_pyplot" href="#ProRF.@show_pyplot"><code>ProRF.@show_pyplot</code></a> — <span class="docstring-category">Macro</span></header><section><div><p>When <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> is on, execute <code>display(gcf())</code> or <a href="#ProRF._julia_interactive"><code>_julia_interactive</code></a> is off, execute <code>show()</code> and wait until the user inputs enter.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/Chemical118/ProRF.jl/blob/7aa5320f631ecca2416c4aba9499c2209f7a5c84/src/ProRF.jl#L106-L108">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../ranfor/">« Random Forest</a><a class="docs-footer-nextpage" href="../findex/">Index »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.18 on <span class="colophon-date" title="Saturday 4 June 2022 07:20">Saturday 4 June 2022</span>. Using Julia version 1.6.6.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
